src/thunderbots/software/geom/util.cpp

```diff
@@ -0,0 +1,914 @@
+        return std::vector<Point>();
+    }
+
+    return std::vector<Point>{a.start +
+                              (a.start - b.start).cross(b.end - b.start) /
+                                  (b.end - b.start).cross(a.end - a.start) *
+                                  (a.end - a.start)};
+}
+
+// ported code
+Vector lineIntersection(
+    const Vector &a, const Vector &b, const Vector &c, const Vector &d)
+{
+    // TODO figure out why this is asserting
+    // assert(std::abs((d - c).cross(b - a)) > EPS);
+    if (std::fabs((d - c).cross(b - a)) < EPS)
+    {
+        // LOG_WARN(u8"Cross product problem again");
+    }
+
+    return a + (a - c).cross(d - c) / (d - c).cross(b - a) * (b - a);
+}
+
+// TODO: a line intersect that takes segments would be nice
+
+Vector reflect(const Vector &v, const Vector &n)
+{
+    if (n.len() < EPS)
+    {
+        return v;
+    }
+    Vector normal = n.norm();
+    return v - 2 * v.dot(normal) * normal;
+}
+
+Vector reflect(const Vector &a, const Vector &b, const Vector &p)
+{
+    // Make a as origin.
+    // Rotate by 90 degrees, does not matter which direction?
+    Vector n = (b - a).rotate(Angle::quarter());
+    return a + reflect(p - a, n);
+}
+
+// ported code
+Vector calcBlockCone(const Vector &a, const Vector &b, const double &radius)
+{
+    if (a.len() < EPS || b.len() < EPS)
+    {
+    }
+    // unit vector and bisector
+    Vector au = a / a.len();
+    Vector c  = au + b / b.len();
+    // use similar triangle
+    return c * (radius / std::fabs(au.cross(c)));
+}
+
+Vector calcBlockCone(
+    const Vector &a, const Vector &b, const Vector &p, const double &radius)
+{
+    return p + calcBlockCone(a - p, b - p, radius);
+}
+
+// ported code
+Vector calcBlockOtherRay(const Vector &a, const Vector &c, const Vector &g)
+{
+    return reflect(c - a, g - c);  // this, and the next two instances, were
+                                   // changed from a - c since reflect() was
+                                   // fixed
+}
+
+// ported code
+bool goalieBlockGoalPost(
+    const Vector &a, const Vector &b, const Vector &c, const Vector &g)
+{
+    Vector R = reflect(c - a, g - c);
+    return fabs(R.cross(b - c)) < EPS;
+}
+
+// ported code
+Vector calcBlockConeDefender(
+    const Vector &a, const Vector &b, const Vector &c, const Vector &g, const double &r)
+{
+    Vector R = reflect(c - a, g - c);
+    // std::cout << (R + c) << std::endl;
+    return calcBlockCone(R + c, b, c, r);
+}
+
+// ported cm code below
+
+double offsetToLine(Vector x0, Vector x1, Vector p)
+{
+    Vector n;
+
+    // get normal to line
+    n = (x1 - x0).perp().norm();
+
+    return n.dot(p - x0);
+}
+
+double offsetAlongLine(Vector x0, Vector x1, Vector p)
+{
+    Vector n, v;
+
+    // get normal to line
+    n = x1 - x0;
+    n = n.norm();
+
+    v = p - x0;
+
+    return n.dot(v);
+}
+
+Vector segmentNearLine(Vector a0, Vector a1, Vector b0, Vector b1)
```

Comment:
no idea what this does

Reply:
nearest point on line segment to line

Reply:
Ah, missed the comment in the header  :P